### DISCUSSION TITLE
Is it possible to bootstrap Cyclone from source without Cyclone?

### DISCUSSION THREAD

**Discussion by Zambito1** [justinethier/cyclone#499] [General]
I'd like to package Cyclone for GNU Guix, but I asked about the bootstrapping requirements on the Guix mailing list, and this was the repsonse: https://mail.gnu.org/archive/html/guix-devel/2023-01/msg00298.html

It seems like using the machine generated C is frowned upon if not unacceptable as a bootstrapping source for Guix. So I'm wondering: is there an early version of Cyclone that can be built using only a C compiler? Or a C compiler in conjunction with another Scheme implementation? This could likely be used to build more recent releases all the way up to the most recent one.

On a related note, it seems like most of the compiler is written in Scheme. I wonder how hard it would be to actually make it so the compiler can be built using any R7RS implementation. The biggest issue to me seems like any calls to C from the compiler would need to have a native Scheme implementation, probably dispatched using `cond-expand`. Does that seem like it could be achievable?

**Comment by justinethier** [comment]
Another data point to keep in mind here is that [CHICKEN's Guix package](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/chicken.scm#n50) calls `make install` which builds that project using generated C code. The difference is that CHICKEN bundles all of the generated `.c` files in the same repository alongside its `.scm` source code. Whereas its more obvious with this project since the generated `.c` code is maintained in a separate repository.

My recommendation is to use `cyclone-bootstrap` to build and package for Guix. I see no difference between that approach and what the CHICKEN package is doing.

**Reply by Zambito1** [reply]
Good point :) I will make a package with `cyclone-bootstrap` and see how that goes.

**Comment by justinethier** [comment]
That's a good question about using another Scheme to build (or run!) the compiler. Its probably possible, the biggest hurdle would be any C code called by the compiler (EG: `define-c` declarations) and potentially any implementation-specific details. Though we *should* be OK there, the idea is to be as standards compliant as possible.

**Reply by Zambito1** [reply]
Thinking about this a bit more, I think only the libraries under `(scheme cyclone ...)` need to have alternative definitions for things defined with `define-c`. At that point, by appending this repo to the library path, any other Scheme should be able to run the complier, as it can use its own implementation of standard R7RS and SRFI libraries.

I haven't done much with `define-c` yet. How does the Cyclone interpreter handle `define-c`? I'm curious if it would make more sense to provide a definition of `define-c` that can be useful for other Schemes, or to provide a native Scheme implementation of everything defined using `define-c`.

**Reply by Zambito1** [reply]
It looks like there are only 11 definitions using `define-c` under `(scheme cyclone ...)` if I counted correctly. Providing an alternative implementation of these in Scheme would probably be very reasonable.

**Reply by justinethier** [reply]
Probably these could all be implemented in native Scheme. The majority of them are for a hashset along with three utility functions.

Since `define-c` is just a thin wrapper around C code there is no way to interpret these expressions. They are compile-only.

**Reply by Zambito1** [reply]
I wrapped those as well as the `inline` library forms in `cond-expand`s and I have gotten pretty far with running the compiler from Chibi. Just working through undefined variables as I hit them, but I think I may be able to get something working :D

**Reply by justinethier** [reply]
Sounds promising! :)

**Reply by Zambito1** [reply]
Here is what I have so far https://git.robbyzambito.me/mirror/cyclone.git/tree/?h=portable-cyclone

I've gotten pretty far with being able to run cyclone.scm using Chibi. The issues I'm running into now are the non-standard extensions to R7RS libraries. Working around that by moving the implementations of those extensions to `(scheme cyclone ...)` (keeping the same extensions to the base libraries by just re-exporting instead) so they are more easily accessible to other implementations.
